tests/unit/common/test_generate_table.py

```diff
@@ -0,0 +1,181 @@
+            catalog=self.spec_catalog)
+
+        self.test_dir = tempfile.mkdtemp()
+        spec_fpath = os.path.join(self.test_dir, 'test.yaml')
+        namespace_fpath = os.path.join(self.test_dir, 'test-namespace.yaml')
+        writer.write_spec(dict(groups=[self.dt_spec, self.dt_spec2]), spec_fpath)
+        writer.write_namespace(self.namespace, namespace_fpath)
+        self.namespace_catalog = NamespaceCatalog()
+        hdmf_typemap = get_type_map()
+        self.namespace_catalog.merge(hdmf_typemap.namespace_catalog)
+        self.type_map = TypeMap(self.namespace_catalog)
+        self.type_map.merge(hdmf_typemap)
+        self.type_map.load_namespaces(namespace_fpath)
+        self.manager = BuildManager(self.type_map)
+
+        self.TestTable = self.type_map.get_container_cls(CORE_NAMESPACE, 'TestTable')
+        self.TestDTRTable = self.type_map.get_container_cls(CORE_NAMESPACE, 'TestDTRTable')
+
+    def tearDown(self) -> None:
+        shutil.rmtree(self.test_dir)
+
+    def test_dynamic_table(self):
+
+        assert issubclass(self.TestTable, DynamicTable)
+
+        assert self.TestTable.__columns__[0] == dict(
+            name='my_col',
+            description='a test column'
+        )
+
+    def test_forbids_incorrect_col(self):
+        test_table = self.TestTable(name='test_table', description='my test table')
+
+        with self.assertRaises(ValueError):
+            test_table.add_row(my_col=3.0, indexed_col=[1.0, 3.0], incorrect_col=5)
+
+    def test_dynamic_column(self):
+        test_table = self.TestTable(name='test_table', description='my test table')
+        test_table.add_column('dynamic_column', 'this is a dynamic column')
+        test_table.add_row(
+            my_col=3.0, indexed_col=[1.0, 3.0], dynamic_column=4, optional_col2=.5,
+        )
+        test_table.add_row(
+            my_col=4.0, indexed_col=[2.0, 4.0], dynamic_column=4, optional_col2=.5,
+        )
+
+        np.testing.assert_array_equal(test_table['indexed_col'].target.data, [1., 3., 2., 4.])
+        np.testing.assert_array_equal(test_table['dynamic_column'].data, [4, 4])
+
+    def test_optional_col(self):
+        test_table = self.TestTable(name='test_table', description='my test table')
+        test_table.add_row(my_col=3.0, indexed_col=[1.0, 3.0], optional_col2=.5)
+        test_table.add_row(my_col=4.0, indexed_col=[2.0, 4.0], optional_col2=.5)
+
+    def test_dynamic_table_region(self):
+
+        TestDTRTable = self.TestDTRTable
+
+        @docval({'name': 'ref_col', 'type': int, 'doc': 'references table'},
+                allow_extra=True)
+        def add_row(self, **kwargs):
+            super(TestDTRTable, self).add_row(**kwargs)
+            for arg, table in zip(['ref_col', 'indexed_ref_col'], [test_table, test_table]):
+                col = self[arg].target if isinstance(self[arg], VectorIndex) else self[arg]
+                if col.table is None:
+                    col.table = table
```

Comment:
I think the following, inserted at line 175, would accomplish the same while being easier to read:
```python
test_dtr_table = self.TestDTRTable(name='test_dtr_table', description='my table')
test_dtr_table['ref_col'].table = test_table
test_dtr_table['indexed_ref_col'].target.table = test_table
```

It looks like a dynamically generated DTR is missing the `table` attribute indicating which DynamicTable it points to, so it needs to be added after generation. If so, this should be documented. For now, documentation in the changelog would do, and we can add this as a TODO
item in the extensions documentation.

Reply:
@rly The thing is you can't do that before adding a row. I can put those lines at the end though:

```python
        test_dtr_table = self.TestDTRTable(name='test_dtr_table', description='my table')

        test_dtr_table.add_row(ref_col=0, indexed_ref_col=[0, 1])
        test_dtr_table.add_row(ref_col=0, indexed_ref_col=[0, 1])

        test_dtr_table['ref_col'].table = test_table
        test_dtr_table['indexed_ref_col'].target.table = test_table
```

is that better?

Reply:
Ideally I'd like to specify what the target tables are in the init. Something like:

```python
test_dtr_table = self.TestDTRTable(name='test_dtr_table', description='my table',
                                   target_tables={'ref_col': test_table, 'indexed_ref_col': test_table})
```

Reply:
Yes, that would be ideal and we can do that. I'll give that a try.

Reply:
Implementing the ideal case proved harder than I thought and will require some more work

Reply:
OK, let's hold off on this feature for now so we can get this through. What we have now works quite well for reading. Even though writing could be improves, it's still much better than what we have now.